zkml/src/layers/transformer/softmax.rs

```diff
@@ -99,16 +390,1019 @@ impl<N: Number> OpInfo for Softmax<N> {
+        ))?;
+        let num_vars = ceil_log2(exp_input.len());
+        // We need to work out how many chunks to split the normalisation into to be range checked.
+        let QuantisedSoftmaxData {
+            error_bound,
+            float_temperature,
+            bkm,
+            lut,
+            ..
+        } = self.quant_info().ok_or(anyhow!(
+            "Could not prove Softmax because it had no quantisation data"
+        ))?;
+        let allowable_error = (*error_bound * OUTPUT_SCALE_FACTOR as f32).round() as Element;
+
+        // Now we construct the polynomials used in the lookups
+        // To do this we need the size of the last dimension
+        let final_dim_size = *step_data.outputs.outputs()[0]
+            .get_shape()
+            .last()
+            .ok_or(anyhow!("Softmax output tensor did not have a shape"))?;
+        let normalisation_lookup = exp_output
+            .chunks(final_dim_size)
+            .map(|chunk| chunk.iter().sum::<Element>())
+            .collect::<Vec<Element>>();
+
+        let merged_range_check = low_range_check
+            .iter()
+            .chain(high_range_check.iter())
+            .copied()
+            .collect::<Vec<Element>>();
+        let merged_softmax = exp_input
+            .iter()
+            .zip(exp_output.iter())
+            .map(|(input, output)| input + output * COLUMN_SEPARATOR)
+            .collect::<Vec<Element>>();
+
+        // Make the commitments to the exp lookup
+        let (exp_commits, exp_evals): (
+            Vec<(PCS::CommitmentWithWitness, DenseMultilinearExtension<E>)>,
+            Vec<Vec<E::BaseField>>,
+        ) = [exp_input, exp_output]
+            .into_par_iter()
+            .map(|vals| {
+                let evaluations = vals
+                    .into_iter()
+                    .map(|v| {
+                        let f: E = v.to_field();
+                        f.as_bases()[0]
+                    })
+                    .collect::<Vec<E::BaseField>>();
+                let mle =
+                    DenseMultilinearExtension::<E>::from_evaluations_slice(num_vars, &evaluations);
+                let commit = ctx.commitment_ctx.commit(&mle)?;
+                Ok(((commit, mle), evaluations))
+            })
+            .collect::<Result<Vec<_>, anyhow::Error>>()?
+            .into_iter()
+            .unzip();
+
+        // Make the commitments to the range checks
+        let (range_commits, range_evals): (
+            Vec<(PCS::CommitmentWithWitness, DenseMultilinearExtension<E>)>,
+            Vec<Vec<E::BaseField>>,
+        ) = [low_range_check, high_range_check]
+            .into_par_iter()
+            .map(|vals| {
+                let evaluations = vals
+                    .into_iter()
+                    .map(|v| {
+                        let f: E = v.to_field();
+                        f.as_bases()[0]
+                    })
+                    .collect::<Vec<E::BaseField>>();
+                let mle =
```

Comment:
It's exactly the same computation - worth doing in parallel for the four MLE no ?